src/main/kotlin/org/rust/lang/core/types/infer/TypeInference.kt

```diff
@@ -0,0 +1,434 @@
+/*
+ * Use of this source code is governed by the MIT license that can be
+ * found in the LICENSE file.
+ */
+
+package org.rust.lang.core.types.infer
+
+import com.intellij.psi.PsiElement
+import org.rust.ide.utils.isNullOrEmpty
+import org.rust.lang.core.psi.*
+import org.rust.lang.core.psi.ext.*
+import org.rust.lang.core.resolve.*
+import org.rust.lang.core.resolve.ref.resolveFieldExprReferenceWithReceiverType
+import org.rust.lang.core.resolve.ref.resolveMethodCallReferenceWithReceiverType
+import org.rust.lang.core.types.ty.*
+import org.rust.lang.core.types.type
+import org.rust.utils.toIntOrNull
+
+fun inferTypesIn(fn: RsFunction): RsInferenceContext {
+    val ctx = RsInferenceContext()
+    TypeInferenceRecursionGuard.guarded { ctx.infer(fn) }
+    return ctx
+}
+
+class RsInferenceContext {
```

Comment:
Not sure that we need the `Rs` prefix here: it is usually used to avoid clashes with classes in the platform. Or do we already have some `InferenceContext` in classpath?